qt/aqt/progress.py

```diff
@@ -223,25 +224,49 @@ def update(
             self._win.form.progressBar.setValue(self._counter)
 
     def finish(self) -> None:
-        # we must latch the levels update and perform it after cleanup has occured or we expose ourselves to a race
-        # condition where a second progress could see levels == 0 and wrongly assume everything is in a clean state
-        next_levels = self._levels - 1
-        next_levels = max(0, next_levels)
-        if next_levels == 0:
-            if self._win:
-                self._closeWin()
-            if self._busy_cursor_timer:
-                self._busy_cursor_timer.stop()
-                self._busy_cursor_timer = None
-            self._restore_cursor()
-            if self._show_timer:
-                self._show_timer.stop()
-                self._show_timer = None
-        if self._backend_timer:
-            self._backend_timer.stop()
-            self._backend_timer.deleteLater()
-            self._backend_timer = None
-        self._levels = next_levels
+        def do_window_cleanup(future: Future | None = None):
+            # this method can be called in an async fashion from taskman where a future
+            # is passed or in synchronous manner from the main thread
+            if future is not None:
+                future.result()
+
+            next_levels = self._levels - 1
+            next_levels = max(0, next_levels)
+            if next_levels == 0:
+                if self._win:
+                    self._closeWin()
+                if self._busy_cursor_timer:
+                    self._busy_cursor_timer.stop()
+                    self._busy_cursor_timer = None
+                self._restore_cursor()
+                if self._show_timer:
+                    self._show_timer.stop()
+                    self._show_timer = None
+            if self._backend_timer:
+                self._backend_timer.stop()
+                self._backend_timer.deleteLater()
+                self._backend_timer = None
+            self._levels = next_levels
+
+        # if the window is not currently shown, we can do cleanup immediately, if it is
+        # currently shown then we need to give the window system a half-second to
+        # present the window before we close it again - fixes progress window getting
+        # stuck, especially on ubuntu 16.10+
+        # NOTE: we must not yield control if the window is not shown since we don't want
+        # to expose ourselves to the possibility of something showing the window in the
+        # meantime
+        if self._shown:
+            elapsed_time = time.monotonic() - self._shown
+            if (time_to_wait := 0.5 - elapsed_time) > 0:
+                self.mw.taskman.run_in_background(
+                    lambda time_to_wait=time_to_wait: time.sleep(time_to_wait),
+                    do_window_cleanup,
+                    uses_collection=False,
+                )
+            else:
+                do_window_cleanup()
+        else:
+            do_window_cleanup()
```

Comment:
This logic could be slightly simplified by calculating the time to sleep, and ensuring it's >= 0. That way the two else paths just becomes delays of 0, meaning they'll get run by Qt in the next main loop invocation.

Reply:
Hmm this one is a little tricky. If `self._shown == 0` then we don't want something to sneak in and make it shown before we get invoked at the back of the timer callback queue, so we have to handle it then and there. We could remove the inner `else` if we max the time to sleep with 0 but that I feel like that obscures the intent of the code.

Unless maybe I am missing your meaning?

Reply:
One alternative would be to set time_to_wait at the top, and then immediately cleanup if not _shown or delay <= 0. It's a minor point in any case, so no strong preference on my end. :-)

Reply:
Okay I see! Thanks for pointing that out. I also have no strong preference and won't read it as many times as you will in the future, so I'll reduce the nesting with the or.